src/hooks/transactions/useTransactionsStatus.tsx

```diff
@@ -0,0 +1,112 @@
+import { handleTransactionUpdates } from '../../util/transactionUtils';
+
+import { useSourceTarget } from '../../contexts/SourceTargetContextProvider';
+import { useApiCallsContext } from '../../contexts/ApiCallsContextProvider';
+import useLaneId from '../chain/useLaneId';
+import { useEffect } from 'react';
+import { TransactionActionCreators } from '../../actions/transactionActions';
+import isEqual from 'lodash/isEqual';
+
+export default function useTransactionsStatus(transactions: TransactionStatusType[], dispatchTransaction: Function) {
+  const { dispatchMessage } = useUpdateMessageContext();
+  const { getValuesByChain } = useChainGetters();
+
+  const subscriptions = useSubscriptionsContext();
+
+  const {
+    sourceChainDetails: { chain: currentSourceChain }
+  } = useSourceTarget();
+  const apiCalls = useApiCallsContext();
+  const laneId = useLaneId();
+
+  useEffect(() => {
+    const getTransactionStatus = async () => {
+      const updatedTransactions = await Promise.all(
+        transactions.map(async (transaction: TransactionStatusType) => {
+          if (transaction.status === TransactionStatusEnum.COMPLETED || transaction.evaluating) {
+            return transaction;
+          }
+          dispatchTransaction(
+            TransactionActionCreators.updateTransactionStatus(
+              {
+                evaluating: true
+              },
+              transaction.id
+            )
+          );
+          const { sourceChain, targetChain } = transaction;
+          const { api: targetApi } = getValuesByChain(targetChain);
+          const { sourceRole, targetRole } = getChainSubscriptionsKey({
+            currentSourceChain,
+            sourceChain
+          });
+
+          const sourceSubscriptions = subscriptions[sourceRole];
+          const targetSubscriptions = subscriptions[targetRole];
+
+          const updatedTransaction = await handleTransactionUpdates({
+            transaction,
+            sourceSubscriptions,
+            targetSubscriptions,
+            apiCalls: {
+              targetApi,
+              ...apiCalls
+            },
+            laneId,
+            dispatchMessage
+          });
+
+          if (isEqual(transaction, updatedTransaction)) {
+            dispatchTransaction(
+              TransactionActionCreators.updateTransactionStatus(
+                {
+                  evaluating: false
+                },
+                transaction.id
+              )
+            );
+          }
+
+          return updatedTransaction;
+        })
+      );
+
+      if (!isEqual(transactions, updatedTransactions)) {
+        dispatchTransaction(TransactionActionCreators.updateTransactionsStatus(updatedTransactions));
+      }
+    };
+
+    getTransactionStatus();
+  }, [
+    apiCalls,
+    currentSourceChain,
+    dispatchMessage,
+    dispatchTransaction,
+    getValuesByChain,
+    laneId,
+    subscriptions,
+    transactions
```

Comment:
This will keep re-evaluating the same transactions every time we just add/change anything in the array unless it's either complete or already evaluating. Isn't there a way to be a bit more fine-grained about it?

Reply:
Ideally the evaluation function should be executed mostly when there is an update on the subscriptions, so i'm wondering if i can catch the previous values of subscriptions ( using `usePrevious` ) and only run the function when the values are different. If we implement this check, we would be omitting the additional execution because of the previous transaction array update. What do you think?

Reply:
But you would need to do that for each transaction separately? I don't know how much extra complexity this would introduce - I'm fine with the current code as-is, was just wondering if there aren't maybe any more fine-grained ways. Ideally we would like only the transaction that actually is affected to be re-calculated, not all of them, but this shouldn't be an issue if we keep the transaction count reasonable.

Reply:
> But you would need to do that for each transaction separately?

No. My idea is to not trigger the function `getTransactionStatus()` until a new subscription value comes in: 

```
const subscriptions = useSubscriptions();
const prevSubs = usePrevious(subscriptions);
...
useEffect( ()=>
...
const subChanged = subscriptions!== prevSubs;
if (!evaluatingTransactions && transactionsInProgress && subChanged) {
      try {
        getTransactionStatus();
      } catch (e) {
```

> Ideally we would like only the transaction that actually is affected to be re-calculated, not all of them.

As the evaluation process mainly resides on updated subscribed values i think there is no scenario when one transaction should be evaluated and the other does not.  So as soon as a new subscribed value arrives all transaction in progress status should be re evaluated.  The exception to this rule is when the transactions are completed , which is a case that is handled already.

Should we go with the subscriptions check? what do you think?

Reply:
Well ... i have added the `const subChanged = subscriptions!== prevSubs;` and i have to say that is working great.
It reduces A LOT unnecessary the number of calls and it is only executed when new values are coming from subscriptions, which is an expected behavior.